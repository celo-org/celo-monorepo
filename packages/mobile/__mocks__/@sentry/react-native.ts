const Sentry = {
  config: () => ({ install: jest.fn() }),
  setTagsContext: jest.fn(),
  captureException: jest.fn(),
}

module.exports = Sentry
