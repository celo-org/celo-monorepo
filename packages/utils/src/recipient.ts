export enum RecipientKind {
  MobileNumber = 'mobileNumber',
  Contact = 'contact',
  QrCode = 'QrCode',
  Address = 'Address',
}
