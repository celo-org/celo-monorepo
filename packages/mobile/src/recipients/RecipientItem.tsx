import ContactCircle from '@celo/react-components/components/ContactCircle'
import colors from '@celo/react-components/styles/colors'
import { fontStyles } from '@celo/react-components/styles/fonts'
import * as React from 'react'
import { Image, StyleSheet, Text, TouchableHighlight, View } from 'react-native'
import { unknownUserIcon } from 'src/images/Images'
import { getRecipientThumbnail, Recipient, RecipientKind } from 'src/recipients/recipient'

interface Props {
  recipient: Recipient
  onSelectRecipient(recipient: Recipient): void
}

class RecipientItem extends React.PureComponent<Props> {
  onPress = () => {
    this.props.onSelectRecipient(this.props.recipient)
  }

  isUnknown = (recipientKind: RecipientKind) => {
    return recipientKind === RecipientKind.Address || recipientKind === RecipientKind.MobileNumber
  }

  render() {
    const { recipient } = this.props

    return (
      <TouchableHighlight onPress={this.onPress} underlayColor={colors.altDarkBg}>
        <View style={style.row}>
          <ContactCircle
            style={style.avatar}
            name={recipient.displayName}
            thumbnailPath={getRecipientThumbnail(recipient)}
            address={recipient.address}
            size={40}
          >
            <Image source={unknownUserIcon} style={style.image} />
          </ContactCircle>
          <View style={style.nameContainer}>
            <Text
              numberOfLines={1}
              ellipsizeMode={'tail'}
              style={[fontStyles.bodySmallSemiBold, style.name]}
            >
              {recipient.displayName}
            </Text>
          </View>
          <Text style={[fontStyles.bodySmallSemiBold, fontStyles.light, style.phone]}>
            {recipient.displayId}
          </Text>
        </View>
      </TouchableHighlight>
    )
  }
}

const style = StyleSheet.create({
  row: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    height: 50,
    paddingHorizontal: 10,
    flex: 1,
  },
  avatar: {
    marginRight: 10,
  },
  nameContainer: {
    flex: 1,
    flexDirection: 'row',
    paddingRight: 5,
  },
  name: {
    lineHeight: 41,
  },
  invite: {
    color: colors.celoGreen,
    lineHeight: 41,
    alignSelf: 'center',
    paddingHorizontal: 10,
  },
  image: {
    height: 40,
    width: 40,
    justifyContent: 'center',
    alignItems: 'center',
  },
  phone: {
    textAlign: 'right',
    lineHeight: 41,
  },
})

export default RecipientItem
