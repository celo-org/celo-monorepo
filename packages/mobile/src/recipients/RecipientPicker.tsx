import Button, { BtnTypes } from '@celo/react-components/components/Button'
import SectionHead from '@celo/react-components/components/SectionHead'
import TextInput, { TextInputProps } from '@celo/react-components/components/TextInput'
import withTextInputLabeling from '@celo/react-components/components/WithTextInputLabeling'
import withTextInputPasteAware from '@celo/react-components/components/WithTextInputPasteAware'
import ForwardChevron from '@celo/react-components/icons/ForwardChevron'
import QRCode from '@celo/react-components/icons/QRCode'
import colors from '@celo/react-components/styles/colors'
import { fontStyles } from '@celo/react-components/styles/fonts'
import { parsePhoneNumber } from '@celo/utils/src/phoneNumbers'
import { isValidAddress } from '@celo/utils/src/signatureUtils'
import { TranslationFunction } from 'i18next'
import * as React from 'react'
import { withNamespaces, WithNamespaces } from 'react-i18next'
import {
  ListRenderItemInfo,
  SectionList,
  SectionListData,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from 'react-native'
import { connect } from 'react-redux'
import { componentWithAnalytics } from 'src/analytics/wrapper'
import { Namespaces } from 'src/i18n'
import Search from 'src/icons/Search'
import { AddressToE164NumberType } from 'src/identity/reducer'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import {
  getRecipientFromAddress,
  NumberToRecipient,
  Recipient,
  RecipientKind,
  RecipientWithAddress,
  RecipientWithMobileNumber,
} from 'src/recipients/recipient'
import RecipientItem from 'src/recipients/RecipientItem'
import { recipientCacheSelector } from 'src/recipients/reducer'
import { RootState } from 'src/redux/reducers'
import DisconnectBanner from 'src/shared/DisconnectBanner'
import { requestContactsPermission } from 'src/utils/androidPermissions'
import Logger from 'src/utils/Logger'
import { assertUnreachable } from 'src/utils/typescript'

const RecipientSearchInput = withTextInputPasteAware(
  withTextInputLabeling<TextInputProps>(TextInput)
)

const goToQrCodeScreen = () => {
  navigate(Screens.QRScanner)
}

const QRCodeCTA = ({ t }: { t: TranslationFunction }) => (
  <TouchableOpacity onPress={goToQrCodeScreen} style={style.qrcodeRow}>
    <View style={style.qrcodeIconLeft}>
      <QRCode />
    </View>
    <View style={style.qrcodeTextContainer}>
      <Text style={[fontStyles.bodySmallSemiBold, style.qrcodeText]}> {t('scanCode')} </Text>
      <Text style={[fontStyles.bodySmall, style.qrcodeText]}>{t('toSentOrRequestPayment')}</Text>
    </View>
    <ForwardChevron height={15} />
  </TouchableOpacity>
)

interface Section {
  key: string
  data: Recipient[]
}

interface Props {
  testID?: string
  showQRCode: boolean
  searchQuery: string
  sections: Section[]
  defaultCountryCode: string
  hasAcceptedContactPermission: boolean
  onSelectRecipient(recipient: Recipient): void
  onSearchQueryChanged(searchQuery: string): void
  onPermissionsAccepted(): void
}

interface StateProps {
  addressToE164Number: AddressToE164NumberType
  recipientCache: NumberToRecipient
}

type RecipientProps = Props & WithNamespaces & StateProps

const mapStateToProps = (state: RootState): StateProps => ({
  addressToE164Number: state.identity.addressToE164Number,
  recipientCache: recipientCacheSelector(state),
})

export class RecipientPicker extends React.Component<RecipientProps> {
  renderItem = ({ item, index }: ListRenderItemInfo<Recipient>) => (
    <RecipientItem recipient={item} onSelectRecipient={this.props.onSelectRecipient} />
  )

  renderSectionHeader = (info: { section: SectionListData<Section> }) => (
    <SectionHead text={info.section.key as string} />
  )

  keyExtractor = (item: Recipient, index: number) => {
    switch (item.kind) {
      case RecipientKind.Contact:
        return item.contactId + item.phoneNumberLabel + index
      case RecipientKind.MobileNumber:
        return item.e164PhoneNumber + index
      case RecipientKind.QrCode:
        return item.address + index
      case RecipientKind.Address:
        return item.address + index
      default:
        Logger.error('RecipientPicker', 'Unsupported recipient kind', item)
        throw assertUnreachable(item)
    }
  }

  renderItemSeparator = () => <View style={style.separator} />

  renderFooter = () => (
    <>
      {this.renderItemSeparator()}
      <Text style={[fontStyles.subSmall, style.footer]}>{this.props.t('searchFriends')}</Text>
    </>
  )

  renderEmptyView = (
    addressToE164Number: AddressToE164NumberType,
    recipientCache: NumberToRecipient
  ) => {
    const parsedNumber = parsePhoneNumber(this.props.searchQuery, this.props.defaultCountryCode)
    if (parsedNumber) {
      return this.renderSendToPhoneNumber(parsedNumber.displayNumber, parsedNumber.e164Number)
    }
    if (isValidAddress(this.props.searchQuery)) {
      return this.renderSendToAddress()
    }
    return this.renderNoContentEmptyView()
  }

  renderNoContentEmptyView = () => (
    <View style={style.emptyView}>
      {this.props.searchQuery !== '' ? (
        <>
          <View style={style.emptyViewBody}>
            <Text style={fontStyles.body}>{this.props.t('noResultsFor')}</Text>
            <Text style={[fontStyles.body, style.emptyViewBodyDark]}>
              {` "${this.props.searchQuery}"`}
            </Text>
          </View>
          <Text style={[fontStyles.subSmall, style.footer]}>
            {this.props.t('searchForSomeone')}
          </Text>
        </>
      ) : (
        <View style={style.emptyViewBody}>
          <Text style={fontStyles.body}>{this.props.t('noContacts')}</Text>
        </View>
      )}
    </View>
  )

  renderSendToPhoneNumber = (displayId: string, e164PhoneNumber: string) => {
    const { t, onSelectRecipient } = this.props
    const recipient: RecipientWithMobileNumber = {
      kind: RecipientKind.MobileNumber,
      displayName: t('mobileNumber'),
      displayId,
      e164PhoneNumber,
    }
    return (
      <>
        <RecipientItem recipient={recipient} onSelectRecipient={onSelectRecipient} />
        {this.renderItemSeparator()}
      </>
    )
  }

  renderRequestContactPermission = () => {
    return (
      <>
        {!this.props.hasAcceptedContactPermission && (
          <Button
            text={this.props.t('askForContactsPermissionAction')}
            style={style.button}
            onPress={this.requestContactsPermission}
            standard={true}
            type={BtnTypes.SECONDARY}
          />
        )}
      </>
    )
  }

  requestContactsPermission = async () => {
    const granted = await requestContactsPermission()

    if (granted) {
      this.props.onPermissionsAccepted()
    }
  }

  renderSendToAddress = () => {
    const { t, searchQuery, addressToE164Number, recipientCache, onSelectRecipient } = this.props
    const existingContact = getRecipientFromAddress(
      searchQuery,
      addressToE164Number,
      recipientCache
    )
    if (existingContact) {
      return (
        <>
          <RecipientItem recipient={existingContact} onSelectRecipient={onSelectRecipient} />
          {this.renderItemSeparator()}
        </>
      )
    } else {
      const recipient: RecipientWithAddress = {
        kind: RecipientKind.Address,
        displayName: t('walletAddress'),
        displayId: searchQuery.substring(2, 17) + '...',
        address: searchQuery,
      }

      return (
        <>
          <RecipientItem recipient={recipient} onSelectRecipient={onSelectRecipient} />
          {this.renderItemSeparator()}
        </>
      )
    }
  }

  render() {
    const { sections, t, addressToE164Number, recipientCache } = this.props
    const showFooter = sections.length > 0

    return (
      <View style={style.body} testID={this.props.testID}>
        <DisconnectBanner />
        <RecipientSearchInput
          placeholder={t('nameOrPhoneNumber')}
          value={this.props.searchQuery}
          onChangeText={this.props.onSearchQueryChanged}
          icon={<Search />}
          style={style.textInput}
          shouldShowClipboard={isValidAddress}
        />
        {this.props.showQRCode && <QRCodeCTA t={t} />}
        <SectionList
          renderItem={this.renderItem}
          renderSectionHeader={this.renderSectionHeader}
          sections={sections}
          ItemSeparatorComponent={this.renderItemSeparator}
          ListFooterComponent={showFooter ? this.renderFooter : undefined}
          ListEmptyComponent={this.renderEmptyView(addressToE164Number, recipientCache)}
          keyExtractor={this.keyExtractor}
          initialNumToRender={30}
          keyboardShouldPersistTaps="handled"
        />
        {this.renderRequestContactPermission()}
      </View>
    )
  }
}

const style = StyleSheet.create({
  body: {
    flex: 1,
  },
  button: {
    marginTop: 30,
    paddingHorizontal: 20,
    paddingVertical: 5,
  },
  textInput: {
    alignSelf: 'center',
    color: colors.dark,
    height: 54,
    marginHorizontal: 8,
  },
  separator: {
    backgroundColor: colors.darkLightest,
    height: 1,
    marginLeft: 60,
  },
  footer: {
    textAlign: 'center',
    paddingVertical: 20,
    paddingHorizontal: 50,
  },
  emptyView: {
    paddingHorizontal: 50,
    justifyContent: 'center',
  },
  emptyViewBody: {
    flexDirection: 'row',
    justifyContent: 'center',
    paddingVertical: 20,
    textAlign: 'center',
  },
  emptyViewBodyDark: {
    color: colors.dark,
  },
  emptyViewBodySmall: {
    justifyContent: 'center',
    textAlign: 'center',
  },
  qrcodeRow: {
    padding: 5,
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    marginVertical: 5,
    marginHorizontal: 10,
  },
  qrcodeIconLeft: {
    borderWidth: 1,
    borderRadius: 15,
    borderColor: colors.celoGreen,
    padding: 4,
  },
  qrcodeTextContainer: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'flex-start',
    textAlignVertical: 'center',
    padding: 3,
  },
  qrcodeText: {
    alignSelf: 'center',
    lineHeight: 30,
  },
})

export default componentWithAnalytics(
  connect(
    mapStateToProps,
    {}
  )(withNamespaces(Namespaces.sendFlow7)(RecipientPicker))
)
