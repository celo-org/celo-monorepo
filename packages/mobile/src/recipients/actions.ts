import { NumberToRecipient } from 'src/recipients/recipient'

export enum Actions {
  SET_RECIPIENT_CACHE = 'SEND/SET_RECIPIENT_CACHE',
}

export interface SetRecipientCacheAction {
  type: Actions.SET_RECIPIENT_CACHE
  recipients: NumberToRecipient
}

export type ActionTypes = SetRecipientCacheAction

export const setRecipientCache = (recipients: NumberToRecipient): SetRecipientCacheAction => ({
  type: Actions.SET_RECIPIENT_CACHE,
  recipients,
})
