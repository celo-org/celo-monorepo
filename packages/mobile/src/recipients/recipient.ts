import { parsePhoneNumber } from '@celo/utils/src/phoneNumbers'
import * as fuzzysort from 'fuzzysort'
import { MinimalContact } from 'react-native-contacts'
import {
  getAddressFromPhoneNumber,
  getVerificationStatusFromPhoneNumber,
  VerificationStatus,
} from 'src/identity/contactMapping'
import { AddressToE164NumberType, E164NumberToAddressType } from 'src/identity/reducer'
import Logger from 'src/utils/Logger'

const TAG = 'recipients/recipient'

export enum RecipientKind {
  MobileNumber = 'MobileNumber',
  Contact = 'Contact',
  QrCode = 'QrCode',
  Address = 'Address',
}

export type Recipient =
  | RecipientWithMobileNumber
  | RecipientWithContact
  | RecipientWithQrCode
  | RecipientWithAddress

interface IRecipient {
  kind: RecipientKind
  address?: string
  displayName: string
  displayId?: string
  e164PhoneNumber?: string
}

export interface RecipientWithMobileNumber extends IRecipient {
  kind: RecipientKind.MobileNumber
  e164PhoneNumber: string // MobileNumber recipients always have a parsable number
}

export interface RecipientWithContact extends IRecipient {
  kind: RecipientKind.Contact
  contactId: string
  phoneNumberLabel?: string
  thumbnailPath?: string
}

export interface RecipientWithQrCode extends IRecipient {
  kind: RecipientKind.QrCode
  address: string
  phoneNumberLabel?: string
  contactId?: string
  thumbnailPath?: string
}

export interface RecipientWithAddress extends IRecipient {
  kind: RecipientKind.Address
  address: string
}

export interface NumberToRecipient {
  [number: string]: RecipientWithContact
}

export function phoneNumberToRecipient(
  e164Number: string,
  address: string | null,
  recipientCache: NumberToRecipient
): Recipient {
  const recipient = recipientCache[e164Number]
  return recipient
    ? {
        kind: RecipientKind.Contact,
        e164PhoneNumber: e164Number,
        displayId: e164Number,
        address: address || undefined,
        phoneNumberLabel: recipient.phoneNumberLabel,
        displayName: recipient.displayName,
        thumbnailPath: recipient.thumbnailPath,
        contactId: recipient.contactId,
      }
    : {
        kind: RecipientKind.MobileNumber,
        e164PhoneNumber: e164Number,
        displayId: e164Number,
        address: address || undefined,
        displayName: e164Number,
      }
}

/**
 * Transforms contacts into a map of e164Number to recipients based on phone numbers from contacts.
 * If a contact has no phone numbers it won't result in any recipients.
 */
export function contactsToRecipients(
  contacts: MinimalContact[],
  defaultCountryCode: string,
  e164NumberToAddress: E164NumberToAddressType
) {
  try {
    //  We need a map of e164Number to recipients so we can efficiently
    //    update them later as the latest contact mappings arrive from the contact calls.
    const e164NumberToRecipients: NumberToRecipient = {}
    // Recipients without e164Numbers go here instead
    const otherRecipients: NumberToRecipient = {}

    for (const contact of contacts) {
      if (!contact.phoneNumbers || !contact.phoneNumbers.length) {
        // Skip contacts without phone numbers
        continue
      }

      for (const phoneNumber of contact.phoneNumbers) {
        const parsedNumber = parsePhoneNumber(phoneNumber.number, defaultCountryCode)

        if (parsedNumber) {
          if (e164NumberToRecipients[parsedNumber.e164Number]) {
            // Skip duplicate phone numbers
            continue
          }
          e164NumberToRecipients[parsedNumber.e164Number] = {
            kind: RecipientKind.Contact,
            displayName: contact.displayName,
            displayId: parsedNumber.displayNumber,
            e164PhoneNumber: parsedNumber.e164Number,
            phoneNumberLabel: phoneNumber.label,
            address: e164NumberToAddress[parsedNumber.e164Number] || undefined,
            contactId: contact.recordID,
            thumbnailPath: contact.thumbnailPath,
          }
        } else {
          otherRecipients[phoneNumber.number] = {
            kind: RecipientKind.Contact,
            displayName: contact.displayName,
            displayId: phoneNumber.number,
            phoneNumberLabel: phoneNumber.label,
            contactId: contact.recordID,
            thumbnailPath: contact.thumbnailPath,
          }
        }
      }
    }

    return { e164NumberToRecipients, otherRecipients }
  } catch (error) {
    Logger.error(TAG, 'Failed to build recipients cache', error)
    return null
  }
}

export function getAddressFromRecipient(
  recipient: Recipient,
  e164NumberToAddress: E164NumberToAddressType
): string | null | undefined {
  if (recipient.kind === RecipientKind.QrCode || recipient.kind === RecipientKind.Address) {
    return recipient.address
  }

  if (!recipient.e164PhoneNumber) {
    throw new Error('Missing recipient e164Number')
  }

  return getAddressFromPhoneNumber(recipient.e164PhoneNumber, e164NumberToAddress)
}

export function getRecipientFromAddress(
  address: string,
  addressToE164Number: AddressToE164NumberType,
  recipientCache: NumberToRecipient
) {
  const e164PhoneNumber = addressToE164Number[address]
  return e164PhoneNumber ? recipientCache[e164PhoneNumber] : undefined
}

export function getRecipientVerificationStatus(
  recipient: Recipient,
  e164NumberToAddress: E164NumberToAddressType
): VerificationStatus {
  if (recipient.kind === RecipientKind.QrCode || recipient.kind === RecipientKind.Address) {
    return VerificationStatus.VERIFIED
  }

  if (!recipient.e164PhoneNumber) {
    throw new Error('No recipient e164Number found')
  }

  return getVerificationStatusFromPhoneNumber(recipient.e164PhoneNumber, e164NumberToAddress)
}

export function getRecipientThumbnail(recipient?: Recipient) {
  return recipient && recipient.kind === RecipientKind.Contact ? recipient.thumbnailPath : undefined
}

type PreparedRecipient = Recipient & {
  displayPrepared: Fuzzysort.Prepared | undefined
  phonePrepared: Fuzzysort.Prepared | undefined
}

type FuzzyRecipient = Recipient | PreparedRecipient

const SCORE_THRESHOLD = -6000

const fuzzysortOptions = {
  keys: ['displayName', 'e164PhoneNumber'],
  threshold: SCORE_THRESHOLD,
}

const fuzzysortPreparedOptions = {
  keys: ['displayPrepared', 'phonePrepared'],
  threshold: SCORE_THRESHOLD,
}

const fuzzysortToRecipients = (
  fuzzyResults: Fuzzysort.KeysResults<FuzzyRecipient>
): FuzzyRecipient[] => {
  // This is the fastest way to map the 'obj' into a results array
  // https://jsperf.com/set-iterator-vs-foreach/16
  const result = []
  for (let _len = fuzzyResults.length, _key = 0; _key < _len; _key++) {
    result[_key] = fuzzyResults[_key].obj
  }
  return result
}

const nameCompare = (a: FuzzyRecipient, b: FuzzyRecipient) => {
  const nameA = a.displayName.toUpperCase()
  const nameB = b.displayName.toUpperCase()

  let comparison = 0
  if (nameA > nameB) {
    comparison = 1
  } else if (nameA < nameB) {
    comparison = -1
  }
  return comparison
}

const executeFuzzySearch = (
  recipients: FuzzyRecipient[],
  query: string,
  options: Fuzzysort.KeysOptions<FuzzyRecipient>,
  shouldSort?: boolean
): FuzzyRecipient[] => {
  const parsedQuery = query.replace(/[()-\s/\\]/g, '')
  if (parsedQuery === '' || parsedQuery.length < 2) {
    // fuzzysort does not handle empty string query
    if (shouldSort) {
      return recipients.sort(nameCompare)
    } else {
      return recipients
    }
  }

  return fuzzysortToRecipients(fuzzysort.go(parsedQuery, recipients, options))
}

export const filterRecipients = (recipients: Recipient[], query: string, shouldSort?: boolean) => {
  return executeFuzzySearch(recipients, query, fuzzysortOptions, shouldSort)
}

export const filterRecipientFactory = (recipients: Recipient[], shouldSort?: boolean) => {
  const preparedRecipients = recipients.map((r) => ({
    ...r,
    displayPrepared: fuzzysort.prepare(r.displayName),
    phonePrepared: r.e164PhoneNumber ? fuzzysort.prepare(r.e164PhoneNumber) : undefined,
  }))

  return (query: string) => {
    return executeFuzzySearch(preparedRecipients, query, fuzzysortPreparedOptions, shouldSort)
  }
}

// Returns true if two recipients are equivalent
// This isn't trivial because two recipients of diff types (Qr code vs contact)
// could potentially refer to the same recipient
export function areRecipientsEquivalent(recipient1: Recipient, recipient2: Recipient) {
  if (recipient1 === recipient2) {
    return true
  }

  if (
    recipient1.e164PhoneNumber &&
    recipient2.e164PhoneNumber &&
    recipient1.e164PhoneNumber === recipient2.e164PhoneNumber
  ) {
    return true
  }

  if (recipient1.address && recipient2.address && recipient1.address === recipient2.address) {
    return true
  }

  // Todo(Rossy) there's still the case where one recip's e164Number gets resolved to another's address
  // but to detect that we'll need to wire in the mappings and check there too

  return false
}
