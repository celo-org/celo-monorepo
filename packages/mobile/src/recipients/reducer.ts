import { Actions, ActionTypes } from 'src/recipients/actions'
import { NumberToRecipient } from 'src/recipients/recipient'
import { RehydrateAction } from 'src/redux/persist-helper'
import { RootState } from 'src/redux/reducers'

export interface State {
  // RecipientCache contains the processed contact data imported from the
  // phone for a single app session.
  // Think of contacts as raw data and recipients as filtered data
  // No relation to recent recipients, which is in /send/reducer.ts
  recipientCache: NumberToRecipient
}

const initialState = {
  recipientCache: {},
}

export const recipientsReducer = (
  state: State | undefined = initialState,
  action: ActionTypes | RehydrateAction
) => {
  switch (action.type) {
    case Actions.SET_RECIPIENT_CACHE:
      return {
        ...state,
        recipientCache: action.recipients,
      }
    default:
      return state
  }
}

export const recipientCacheSelector = (state: RootState) => state.recipients.recipientCache
