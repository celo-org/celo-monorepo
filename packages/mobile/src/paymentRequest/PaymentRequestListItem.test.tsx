import * as React from 'react'
import 'react-native'
import * as renderer from 'react-test-renderer'
import PaymentRequestListItem from 'src/paymentRequest/PaymentRequestListItem'
import { RecipientKind } from 'src/recipients/recipient'

const commonProps = {
  id: 1,
  amount: '24',
  comment: 'Hey thanks for the loan, Ill pay you back ASAP. LOVE YOU',
  updatePaymentRequestStatus: jest.fn(),
  requester: {
    kind: RecipientKind.MobileNumber,
    e164PhoneNumber: '5126608970',
    displayId: '5126608970',
    address: '0x91623f625e23ac1400',
    displayName: '5126608970',
    contact: undefined,
  },
}

describe('PaymentRequestListItem', () => {
  it('renders correctly', () => {
    // @ts-ignore -- kind is not assignable?
    const tree = renderer.create(<PaymentRequestListItem {...commonProps} />)

    expect(tree).toMatchSnapshot()
  })
})
