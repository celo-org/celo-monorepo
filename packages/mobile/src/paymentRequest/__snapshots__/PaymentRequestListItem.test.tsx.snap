// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaymentRequestListItem renders correctly 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
          "padding": 16,
          "width": "100%",
        },
        false,
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "column",
          "paddingRight": 16,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
              },
              Object {
                "backgroundColor": "#7AD6FE",
                "borderRadius": 15,
                "height": 30,
                "width": 30,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#FFFFFF",
                  "fontFamily": "Hind-Medium",
                  "fontSize": 16,
                },
                Object {
                  "fontSize": 15,
                },
              ]
            }
          >
            5
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "justifyContent": "space-between",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#2E3338",
            "fontFamily": "Hind-Bold",
            "fontSize": 14,
            "lineHeight": 18,
          }
        }
      >
        5126608970
      </Text>
      <View>
        <View
          style={
            Object {
              "justifyContent": "space-between",
              "minHeight": 70,
            }
          }
        >
          <View
            style={Object {}}
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#81868B",
                    "fontFamily": "Hind-Regular",
                    "fontSize": 14,
                    "lineHeight": 20,
                  },
                  Object {
                    "paddingTop": 16,
                  },
                ]
              }
            >
              Hey thanks for the loan, Ill pay you back ASAP. LOVE YOU
            </Text>
          </View>
          <View
            style={
              Object {
                "flexDirection": "row",
                "marginTop": 15,
              }
            }
          >
            <View
              accessible={true}
              isTVSelectable={true}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#2E3338",
                      "fontFamily": "Hind-Regular",
                      "fontSize": 14,
                      "lineHeight": 18,
                    },
                    Object {
                      "color": "#42D689",
                      "fontFamily": "Hind-Medium",
                      "fontSize": 14,
                      "lineHeight": 18,
                    },
                    Object {
                      "paddingEnd": 15,
                    },
                  ]
                }
              >
                pay
              </Text>
            </View>
            <View
              accessible={true}
              isTVSelectable={true}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#2E3338",
                      "fontFamily": "Hind-Regular",
                      "fontSize": 14,
                      "lineHeight": 18,
                    },
                    Object {
                      "color": "#42D689",
                      "fontFamily": "Hind-Medium",
                      "fontSize": 14,
                      "lineHeight": 18,
                    },
                    Object {
                      "paddingEnd": 15,
                    },
                  ]
                }
              >
                decline
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "paddingLeft": 16,
        }
      }
    >
      <View>
        <Text
          style={
            Array [
              Object {
                "color": "#2E3338",
                "fontFamily": "Hind-SemiBold",
                "fontSize": 14,
                "lineHeight": 18,
              },
              Object {
                "color": "#81868B",
              },
            ]
          }
        >
          $24
        </Text>
      </View>
    </View>
  </View>
</View>
`;
